### ISSUE TITLE
keybindings should not work in normal mode by default

### ISSUE THREAD

**Issue by seanmcbreen** [Microsoft/vscode#575]
_From @guillermooo on November 8, 2015 13:4_

For Vim emulation, there needs to be a way to capture key strokes and suppress them before they are inserted in the buffer. That is, normal mode in Vim.

In other editors, it's a problem that all key bindings work by default in normal mode. Because plugins are loaded alphabetically, there is then a chance that the plugin to load last may override Vim key bindings, even though that plugin isn't interested in normal mode at all and/or doesn't even work when the editor is in normal mode.

When VS Code enables a variant of normal mode (when? :-)), I think it'd be best if key bindings had to opt-in explicitly to normal mode.

_Copied from original issue: Microsoft/vscode-extensionbuilders#83_

**Comment by seanmcbreen** [comment]
Adding @alexandrudima to this thread as our Key Bindings guru.

**Comment by bgourlie** [comment]
>When VS Code enables a variant of normal mode

Are there plans to make VS Code a moda editor?  I assume the comment was just being hopeful, but that would be nice.

In any event, I would switch from vim to VS Code in a heartbeat if it had halfway decent vim emulation.  I've already started scouring the extension API and came across this ticket when wondering myself how one would implement normal mode (or any non-insert mode, really).  I've submitted a somewhat related ticket [here](https://github.com/Microsoft/vscode/issues/561).

Another thing I would love to see is a way to alter keybindings for some of the built-in widgets, like the command palette and intellisense lists.  In particular, I'd like to be able to configure them to use J and K for moving up and down (or Command-K/J if an autocomplete list).  There is a ton of opportunity to make VS Code an amazing editor for those of us who can't live without modal editing.
